In [1]:
import sys
# Check Python version
if sys.version_info.major == 3 and sys.version_info.minor >= 6:
    print("Hello, World!")
    print(f"Python version: {sys.version}")
else:
    print("Please use Python 3.6 or higher.")

Hello, World!
Python version: 3.12.6 (tags/v3.12.6:a4a2d2b, Sep  6 2024, 20:11:23) [MSC v.1940 64 bit (AMD64)]


In [2]:
# Mutable Data Type Example (List)
my_list = [1, 2, 3]
print("Original List:", my_list)
# Modifying the list (changing contents)
my_list.append(4)
print("Modified List:", my_list)
# Immutable Data Type Example (Tuple)
my_tuple = (1, 2, 3)
print("Original Tuple:", my_tuple)
 #Tryingtomodifythetuple(willraiseanerror)
# my_tuple.append(4)#UncommentingthislinewillraiseanAttributeError

Original List: [1, 2, 3]
Modified List: [1, 2, 3, 4]
Original Tuple: (1, 2, 3)


In [1]:
age= 23#int
pi=3.14#float
temperature=-5.5#float
print("datatypeofvariableage=",type(age))
print("datatypeofvariablepi=",type(pi))
print("datatypeofvariabletemperature=",type(temperature))

datatypeofvariableage= <class 'int'>
datatypeofvariablepi= <class 'float'>
datatypeofvariabletemperature= <class 'float'>


In [5]:
name = "Alice"
# greeting = ’Hello’
address = "123 Main St"
print("data type of the variable name = ",type(name))
# print("data type of the variable greeting = ",type(greeting))
print("data type of the variable address = ",type(address))
# slice only one element
print("The first letter of the name is:", name[0])
print("The last letter of the name is:", name[-1])
# slice a range of elements
print("The second letter to the fourth of the name is:", name[1:4])
print("The first two letters of the name are:", name[:2])
print("Substring starting from the third letter is:", name[2:])

data type of the variable name =  <class 'str'>
data type of the variable address =  <class 'str'>
The first letter of the name is: A
The last letter of the name is: e
The second letter to the fourth of the name is: lic
The first two letters of the name are: Al
Substring starting from the third letter is: ice


In [7]:
list1 = [1, 2, 3, 4]
mixed_list = [12, "Hello", True]
# List is mutable
mixed_list[0] = "Sam"
mixed_list

['Sam', 'Hello', True]

In [9]:
colors = ("red", "green", "yellow", "blue")
print("First element:", colors[0])
print("Last two elements:", colors[2:])
print("Middle two elements:", colors[1:3])
# colors[0] = "purple"
colors # will generate an error as tuple is immutable.

First element: red
Last two elements: ('yellow', 'blue')
Middle two elements: ('green', 'yellow')


('red', 'green', 'yellow', 'blue')

In [6]:
person = {'name': 'john', 'age': 25, 'city': "Kathmandu"}
print(f"Hello my name is {person['name']}, I am {person['age']} years old and I live at {person['city']}.")
print("All Keys:", person.keys())
print("All Values:", list(person.values()))

Hello my name is john, I am 25 years old and I live at Kathmandu.
All Keys: dict_keys(['name', 'age', 'city'])
All Values: ['john', 25, 'Kathmandu']


In [ ]:
num = 10
if num > 0:
    print("Positive Number")
elif num == 0:
    print("Zero")
else:
    print("Negative Number")

Positive Number


In [9]:
x = 5
y = 10
if x > 0 and y < 20:
    print("Both condition are true")
if x > 0 or y > 20:
    print("At least one condition is true")
if not x == 0:
    print("X is not equal to 0")


Both condition are true
At least one condition is true
X is not equal to 0


In [11]:
a = True
b = False

if a and b:
    print("Both a and b are True")
else:
    print("Either a or b is False")

if a or b:
    print("At least one of a or b is True")
else:
    print("Both a and b are False")

if not a:
    print("A is False")
else:
    print("A is True")


Either a or b is False
At least one of a or b is True
A is True


For Loop

In [12]:
Fruits = ["Apple", "Banana", "Mango", "Orange"]

for fruit in Fruits:
    print(fruit)


Apple
Banana
Mango
Orange


While Loop

In [13]:
count = 0
while count < 5:
    print("Count is: ", count)
    count += 1
    

Count is:  0
Count is:  1
Count is:  2
Count is:  3
Count is:  4


Break

In [15]:
Fruits = ["Apple", "Banana", "Cherry", "Mango", "Orange"]
print("Loop -> 1")
for fruit in Fruits:
    print(fruit)
    if fruit == "Mango":
        break

print("Loop -> 2")
for fruit in Fruits:
    if fruit == "Mango":
        continue
    else:
        print(fruit)

Loop -> 1
Apple
Banana
Cherry
Mango
Loop -> 2
Apple
Banana
Cherry
Orange


Function

In [ ]:
def add_binary(a,b):
    binary_sum = bin(a+b)[2:]
    return binary_sum

print(add_binary(2,2))
print(add_binary(51,12))

100
111111


Built-in Functions

In [19]:
print("Hello 12")
length = len("Hello")
print(length)
data_type = type("Hello")
print(data_type)
total = sum([1,3,2,4])
print(total)


Hello 12
5
<class 'str'>
10


Built-in Methods

In [32]:
message = "Hello Samyog"
message_upper = message.upper()
print(message_upper)

Animals = ["Dog", "Cat", "Lion", "Tiger"]
print("Original List:", Animals)
# Append function
Animals.append("Ant")
print("Updated List:", Animals)

# Map
info = {"Name": "Sam", "Age": 20}
age = info.get("Age")
print("Age:",age)

HELLO SAMYOG
Original List: ['Dog', 'Cat', 'Lion', 'Tiger']
Updated List: ['Dog', 'Cat', 'Lion', 'Tiger', 'Ant']
Age: 20


User-Defined Function

In [35]:
def hello123():
    # print hello
    print("Choose the conversation type:")
    print("1. Celsius to fahrenheit")
    print("2. Fahrenheit to Celsius")
    
    choice = int(input("Enter 1 or 2:"))
    if choice == 1:
        celsius = float(input("Enter temperature in Celsius: "))
        fahrenheit = (celsius * 9/5) + 32
        print(f"{celsius} C is equal to {fahrenheit}F")
        return fahrenheit
    elif choice == "2":
        fahrenheit = float(input("Enter Temperature in fahrenheit: "))
        celsius = (fahrenheit - 32) * 5/9
        print(f"{fahrenheit} F is equal to {celsius}C")
        return celsius
    else:
        print("invalid choice. Please enter 1 or 2.")
        return None
    
hello123()

Choose the conversation type:
1. Celsius to fahrenheit
2. Fahrenheit to Celsius
37.0 C is equal to 98.6F


98.6

Global And Local Variables

In [40]:
x = 10 # Global Variable
def print_global():
    print(x)

print_global()

def print_local():
    s = 5
    print(s)
    
print_local()
# print(s)

10
5


In [46]:
# Modify of Global Variable
global_value = 30
def modify_global():
    global global_value
    global_value = 10
    print(global_value)

modify_global()
print(global_value)

10
10


Exception Handling

In [57]:
try:
    num = float(input("Enter a number:"))
    print("Result", 10/num)
except ZeroDivisionError:
    print("Cannot Divide by Zero")
except ValueError:
    print("Only Enter a number")

Only Enter a number


With Finally

In [ ]:
try:
    file = open("example.txt", "r")
    data = file.read()
except:
    print("File not found")
finally:
    file.close()

Raising Custom Exceptions

In [61]:
def check_age(age):
    if age < 0:
        raise ValueError("Age cannot be negative")
    else:
        print("Valid AGE")

try:
    check_age(3)
except ValueError as e:
    print(e)

Valid AGE


Tasks

In [3]:
# Task 1
temperatures = [8.2, 17.4, 14.1, 7.9, 18.0, 13.5, 9.0, 17.8, 13.0, 8.5,
16.5, 12.9, 7.7, 17.2, 13.3, 8.4, 16.7, 14.0, 9.5, 18.3, 13.4, 8.1,
17.9, 14.2, 7.6, 17.0, 12.8, 8.0, 16.8, 13.7, 7.8, 17.5, 13.6, 8.7,
17.1, 13.8, 9.2, 18.1, 13.9, 8.3, 16.4, 12.7, 8.9, 18.2, 13.1, 7.8,
16.6, 12.5]

Cold = []
cold_count = 0
Mild = []
mild_count = 0
Comfortable = []
comfortable_count = 0

for temp in temperatures:
    if (temp < 10):
        cold_count = cold_count+1
        Cold.append(temp)
    elif (temp > 10 and temp > 15):
        mild_count = mild_count+1
        Mild.append(temp)
    else:
        comfortable_count = comfortable_count +1
        Comfortable.append(temp)
        
print("Cold: ", Cold)
print("Mild:", Mild)
print("Comfortable:", Comfortable)

# Task 2
# Total number of value in each
print("Cold Count", cold_count)
print("Mild Count", mild_count)
print("Comfortable Count", comfortable_count)

print("Count using built in function:", len(Cold))
print("Count using built in function:", len(Mild))
print("Count using built in function: ", len(Comfortable))

Cold:  [8.2, 7.9, 9.0, 8.5, 7.7, 8.4, 9.5, 8.1, 7.6, 8.0, 7.8, 8.7, 9.2, 8.3, 8.9, 7.8]
Mild: [17.4, 18.0, 17.8, 16.5, 17.2, 16.7, 18.3, 17.9, 17.0, 16.8, 17.5, 17.1, 18.1, 16.4, 18.2, 16.6]
Comfortable: [14.1, 13.5, 13.0, 12.9, 13.3, 14.0, 13.4, 14.2, 12.8, 13.7, 13.6, 13.8, 13.9, 12.7, 13.1, 12.5]
Cold Count 16
Mild Count 16
Comfortable Count 16
Count using built in function: 16
Count using built in function: 16
Count using built in function:  16


In [5]:
# Task 3
def convertCelToFahrenheit(celsius):
    fahrenheit = celsius * (9/5) + 32
    return fahrenheit

fahrenheit_s = []

for celsius in temperatures:
    fah = convertCelToFahrenheit(celsius)
    fahrenheit_s.append(fah)
    
print("Fahrenheit List: ", fahrenheit_s)
    

Fahrenheit List:  [46.76, 63.31999999999999, 57.379999999999995, 46.22, 64.4, 56.3, 48.2, 64.03999999999999, 55.400000000000006, 47.3, 61.7, 55.22, 45.86, 62.96, 55.94, 47.120000000000005, 62.06, 57.2, 49.1, 64.94, 56.120000000000005, 46.58, 64.22, 57.56, 45.68, 62.6, 55.040000000000006, 46.4, 62.24, 56.66, 46.04, 63.5, 56.480000000000004, 47.66, 62.78, 56.84, 48.56, 64.58000000000001, 57.019999999999996, 46.94, 61.519999999999996, 54.86, 48.019999999999996, 64.75999999999999, 55.58, 46.04, 61.88, 54.5]


# To-do Task

1. Task-1

In [6]:
def meter_to_feet(meter):
    '''
    This function converts the meter to feet by, 
    converting it first to inch then to feet.
    
    it has both print and return value incase i want to use return instead of print.
    
    it takes meter and convert to feet.
    '''
    centimeter = meter*100
    inch = centimeter/2.5
    feet = inch/12
    print(feet)
    return feet

def feet_to_meter(feet):
    '''
    This function converts the feet to meter by,
    converting feet ot inch to centimeter to finally meter
    
    it has both print and return value incase i want to use return instead of print.
 
    It take feet as a parameter and converts it to meter
    '''
    inch = feet*12
    centimeter = inch*2.5
    print(centimeter/100)
    return centimeter/100

def kg_to_pound(kg):
    '''
    This function converts the kg to pound
    it converts the kg to grams and then grams to kg
    
    it has both print and return value incase i want to use return instead of print.

    It takes kg as a parameter and converts it to pound
    '''
    grams = kg*1000
    print(grams/453.59237)
    return grams/453.59237

def pound_to_kg(pound):
    '''
    This function converts the pound to kg
    it converts the pound to grams and then grams to pound
    
    it has both print and return value incase i want to use return instead of print.

    It takes pound as a parameter and converts it to kg
    '''
    grams = pound * 453.59237
    print(grams/1000)
    return grams/1000

def liters_to_gallons(liters):
    '''
    This function converts the liters to gallons
    it converts the liters to gallons by dividing it by 3.78
    
    it has both print and return value incase i want to use return instead of print. 
    
    It takes liters and returns gallons
    '''
    print(liters/3.78)
    return liters/3.78

def gallons_to_liters(gallons):
    '''
    This function converts the gallons to liters
    It converts the gallons to liters by multiplying it by 3.78
    
    it has both print and return value incase i want to use return instead of print.
    
    It takes gallons and returns liters
    '''
    print(gallons*3.78)
    return gallons*3.78

def Length():
    try:
        nested_choice = int(input("Enter (1 to choose meter to feet), AND Enter (2 to choose feet to meter)"))
        if(nested_choice == 1):
            try:
                meter = float(input("Enter the length in meter: "))
                meter_to_feet(meter)
            except ValueError as e:
                print("Error: ", e)
        elif(nested_choice == 2):
            try:
                feet = float(input("Enter the length in feet: "))
                feet_to_meter(feet)
            except ValueError as e:
                print("Error: ",e)
        else:
            print("Please select 1 or 2")
    except ValueError as e:
        print("Error: ", e)

def Width():
    try:
        nested_choice = int(input("Enter (1 to convert Kg to pound) AND (2 to convert pound to Kg)"))
        if(nested_choice == 1):
            try:
                kg = float(input("Enter the kg: "))
                kg_to_pound(kg)
            except ValueError as e:
                print("Error: ", e)
        elif(nested_choice == 2):
            try:
                pound = float(input("Enter in pound: "))
                pound_to_kg(pound)
            except ValueError as e:
                print("Error: ", e)
        else:
            print("Please select 1 ro 2")
    except ValueError as e:
        print("Error", e)
        
def Volume():
    try:
        nested_choice = int(input("Enter (1 to convert liters to gallons) AND (2 to convert gallons to liters)"))
        if(nested_choice == 1):
            try:
                liters = float(input("Enter the volume in liters: "))
                liters_to_gallons(liters)
            except ValueError as e:
                print("Error: ", e)
        elif(nested_choice == 2):
            try:
                gallons = float(input("Enter the volume in gallons: "))
                gallons_to_liters(gallons)
            except ValueError as e:
                print("Error: ", e)
        else:
            print("Enter the choice 1 or 2 only")
    except ValueError as e:
        print("Error: ", e)

def main():
    try:
        choice = int(input("Enter the choice (1 for Length) (2 for Width) and (3 for Volume): "))
        if(choice == 1):
            Length()
        elif(choice == 2):
            Width()
        elif(choice == 3):
            Volume()
        else:
            print("Please Select the 1, 2 or 3")
    except ValueError as e:
        print("Error: ", e)
        
main()


400.0


# Task 2

In [7]:
def find_sum(list):
    '''
    This function take the list as an argument and return its sum value
    '''
    sum = 0
    for num in list:
        sum = sum + num
        
    return sum

def find_average(list):
    '''
    This function take the list as an argument and return its average value
    '''
    sum = sum(list)
    average = sum/len(list)
    return average

def find_maximum(list):
    '''
    This function takes the list as an argument and return its maximum value
    '''
    max_value = list[0]
    for num in list:
        if num > max_value:
            max_value = num
    
    return max_value

def find_minimum(list):
    '''
    This function takes the list as an argument and return its minimum value
    '''
    min_value = list[0]
    for num in list:
        if(num < min_value):
            min_value = num
            
    return min_value

def take_list():
    '''
    This function takes the list from user with space between each number or member
    '''
    try:
        list = input("Enter the lists with space between: ")
        
        if not list.strip():
            print("Error: No value Entered")
            return None
        
        user_list = [int(num) for num in list.split()]
        
        if (len(user_list) == 0):
            print("Error: The list is empty!")
            return 0
        
        return user_list        
    except ValueError as e:
        print("Error: ", e)
    

def user_choice():
    '''
    This function let user to choice between 1,2,3 and 4
    '''
    try:
        choice = int(input("Enter the (1 to find sum), (2 to average), (3 to find maximum) and (4 to find minimum)"))
        if (choice == 1):
            list = take_list()
            sum = find_sum(list)
            print("Sum:", sum)
        elif(choice == 2):
            take_list()
            average = find_average(list)
            print("Average: ", average)
        elif(choice == 3):
            take_list()
            maximum = find_maximum(list)
            print("Maximum: ", maximum)
        elif(choice == 4):
            take_list()
            minimum = find_minimum(list)
            print("Minimum: ", minimum)
        else:
            print("Please choose: 1, 2, 3 and 4 only!")
    except ValueError as e:
        print("Error: ", e)
        
user_choice()

Sum: 15


# 4.1

Question 1: Extract Every Other Element

In [70]:
def extract_every_other(lst):
    for i in range(0, len(lst), 2):
        print(lst[i])

list = [1, 2, 3, 4, 5, 6]
extract_every_other(list)

1
3
5


Question 2: Slice a Sublist

In [84]:
def get_sublist(lst, start, end):
    new_list = []
    for i in range(start, end+1, 1):        
        new_list.append(lst[i])
    return new_list

user_input = [1, 2, 3, 4, 5, 6]
new_list = get_sublist(user_input, 2,4)

print(new_list)


[3, 4, 5]


Question 3: Reverse a List Using Slicing

In [3]:
def reverse_list(lst):
    return lst[::-1]

# Example usage:
input_list = [1, 2, 3, 4, 5]
output_list = reverse_list(input_list)
print(output_list)


[5, 4, 3, 2, 1]


Question 4: Remove the First and Last Elements

In [11]:
def remove_first_last(lst):
    new_list = []
    for i in range(1, len(lst)-1):
        new_list.append(lst[i])
        
    return new_list

user_input = [1, 2, 3, 4, 5]
remove_first_last(user_input)


[2, 3, 4]

Question 6: Extract Elements from the End

In [6]:
def get_last_n(lst, n):
    new_list = []
    lst = reverse_list(lst)
    for i in range(n):
        new_list.append(lst[i])
    return reverse_list(new_list)

user_input = [1, 2, 3, 4, 5]
print(get_last_n(user_input, 2))

[4, 5]


Question 7: Extract Elements in Reverse Order

In [38]:
def reverse_skip(lst):
    new_list = []
    for i in range(len(lst)-2, -1, -2):
        new_list.append(lst[i])
    return new_list

user_input =  [1, 2, 3, 4, 5, 6]

reverse_skip(user_input)

[5, 3, 1]

# 4.3

Question 1: Flatten a Nested List

In [64]:
def flatten_extend(lst):
     flat_list = []
     for item in lst:
        if isinstance(item, list):
            flat_list.extend(flatten_extend(item))
        else:
            flat_list.append(item)

     return flat_list

user_input = [[1, 2], [3, [2 , [4,5], 3],  4], [5]]
print(flatten_extend(user_input))

[1, 2, 3, 2, 4, 5, 3, 4, 5]


Question 2: Accessing Nested List Elements

In [73]:
def access_nested_element(lst, indices):
    for index in indices:
        lst = lst[index]
    return lst

lst = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
indics = [1,2]
print(access_nested_element(lst, indics))

6


Question 3: Sum of All Elements in a Nested List

In [86]:
def sum_nested(lst):
    sum = 0
    for item in lst:
        if isinstance(item, list):
            sum += sum_nested(item)
        else:
            sum += item
    return sum

user_input = [[1, 2], [3, [4, 5]], 6]
print(sum_nested(user_input))

21


Question 4: Remove Specific Element from a Nested List

In [100]:
def remove_element(lst, elem):
    for i in range(len(lst)):
        lst[i] = [x for x in lst[i] if x != elem]
    return lst

lst = [[1, 2], [3, 2], [4, 5]]
elem = 2
result = remove_element(lst, elem)
print(result)

[[1], [3], [4, 5]]


Question 5: Find the Maximum Element in a Nested List

In [4]:
def find_max(lst):
    max_item = float('-inf')
    for item in lst:
        if (isinstance(item, list)):
            max_item = max(max_item, find_max(item))
        else:
            max_item = max(max_item, item)
    return max_item
            
user_input = [[1, 2], [3, [4, 5]], 6]
find_max(user_input)

6

Question 6: Count Occurrences of an Element in a Nested List

In [15]:
def count_occurrences(lst, elem):
    count = 0
    for item in lst:
        if isinstance(item, list):
            for i in item:
                if i == elem:
                    count +=1
        else:
            if item == elem:
                count += 1
    return count

input_lst = [[1, 2], [2, 3], [2, 4], 2]
elem = 2

print(count_occurrences(input_lst, elem))

4


Question 7: Flatten a List of Lists of Lists

In [22]:
def deep_flatten(lst):
    combine_list = []
    for item in lst:
        if isinstance(item, list):
            combine_list.extend(deep_flatten(item))
        else:
            combine_list.append(item)
    return combine_list

user_input = [[[1, 2], [3, 4]], [[5, 6], [7, 8]]]

print(deep_flatten(user_input))

[1, 2, 3, 4, 5, 6, 7, 8]


Question 8: Nested List Average

In [ ]:
def nested_average(lst):
    sum = 0
    count = 0
    for item in lst:
        if isinstance(item, list):
            sub_sum, sub_count = nested_average(item)
            sum += sub_sum
            count += sub_count
        else:
            sum += item
            count += 1
    return sum, count

def find_average(user_input):
    sum, count = nested_average(user_input)    
    return sum/count

user_input = [[1, 2], [3, 4], [5, 6]]
print(find_average(user_input))

3.5


# NumPy

In [8]:
import numpy as np
# initialize an all zero array with size 2 X 3:
zeros_arr = np.zeros((2,3))
print(" A zeros array is \n", zeros_arr, "with dimensions", zeros_arr.shape, "\n")


 A zeros array is 
 [[0. 0. 0.]
 [0. 0. 0.]] with dimensions (2, 3) 



In [38]:
import numpy as np
# Create and display zero, one, and n-dimensional arrays
zero_dim_array = np.array(5)
one_dim_array = np.array([1, 2, 3])
n_dim_array = np.array([[1, 2], [3, 4]])
for arr in [zero_dim_array, one_dim_array, n_dim_array]:
    print(f"Array:\n{arr}\nDimension: {arr.ndim}\nData type: {arr.dtype}\n")

Array:
5
Dimension: 0
Data type: int64

Array:
[1 2 3]
Dimension: 1
Data type: int64

Array:
[[1 2]
 [3 4]]
Dimension: 2
Data type: int64



In [39]:
import numpy as np
# Create an array with shape (2, 3)
array = np.array([[1, 2, 3],
[4, 5, 6]])

# Reshape to (3, 2), keeping 6 elements
reshaped_array = array.reshape(3, 2)
print("Original Shape:", array.shape, "\nReshaped Shape:", reshaped_array.shape)
# Expected Outputs:
# Original Shape: (2, 3)
# Reshaped Shape: (3, 2)


Original Shape: (2, 3) 
Reshaped Shape: (3, 2)


# Exercises

Problem 1. Array Creation

In [1]:
import numpy as np

# 1. Initialize an empty array with size 2x2
empty_array = np.empty((2, 2))
print("Empty array (2x2):\n", empty_array)

# 2. Initialize an all one array with size 4x2
ones_array = np.ones((4, 2))
print("\nAll ones array (4x2):\n", ones_array)

# 3. Return a new array of given shape and type, filled with a fill value
fill_value_array = np.full((3, 3), 7)  
print("\nArray filled with 7:\n", fill_value_array)

# 4. Return a new array of zeros with the same shape and type as a given array
existing_array = np.array([1, 2, 3, 4]) 
zeros_like_array = np.zeros_like(existing_array)
print("\nArray of zeros with the same shape as existing_array:\n", zeros_like_array)

# 5. Return a new array of ones with the same shape and type as a given array
ones_like_array = np.ones_like(existing_array)
print("\nArray of ones with the same shape as existing_array:\n", ones_like_array)

# 6. For an existing list new_list = [1, 2, 3, 4], convert to a numpy array
new_list = [1, 2, 3, 4]
numpy_array = np.array(new_list)
print("\nConverted numpy array:\n", numpy_array)


Empty array (2x2):
 [[4.78656376e-315 0.00000000e+000]
 [4.74575602e-315 4.77539316e-315]]

All ones array (4x2):
 [[1. 1.]
 [1. 1.]
 [1. 1.]
 [1. 1.]]

Array filled with 7:
 [[7 7 7]
 [7 7 7]
 [7 7 7]]

Array of zeros with the same shape as existing_array:
 [0 0 0 0]

Array of ones with the same shape as existing_array:
 [1 1 1 1]

Converted numpy array:
 [1 2 3 4]


Problem 2. Array Manipulation: Numerical Ranges and Array indexing

In [7]:
import numpy as np

# Task 1: Create an array with values ranging from 10 to 49
arr_1 = np.arange(10, 50)
print("Task 1:", arr_1, "\n")

# Task 2: Create a 3X3 matrix with values ranging from 0 to 8
arr_2 = np.arange(9).reshape(3, 3)
print("Task 2:\n", arr_2, "\n")

# Task 3: Create a 3X3 identity matrix
arr_3 = np.eye(3)
print("Task 3:\n", arr_3)

# Task 4: Create a random array of size 30 and find the mean
arr_4 = np.random.random(30)
mean_arr_4 = arr_4.mean()
print("Task 4: Mean of array =", mean_arr_4, "\n")

# Task 5: Create a 10X10 array with random values and find the minimum and maximum values
arr_5 = np.random.random((10, 10))
min_val = arr_5.min()
max_val = arr_5.max()
print("Task 5: Min =", min_val, "Max =", max_val, "\n")

# Task 6: Create a zero array of size 10 and replace the 5th element with 1
arr_6 = np.zeros(10)
arr_6[4] = 1  # Index 4 is the 5th element (0-based index)
print("Task 6:", arr_6, "\n")

# Task 7: Reverse an array arr = [1, 2, 0, 0, 4, 0]
arr_7 = np.array([1, 2, 0, 0, 4, 0])
reversed_arr_7 = arr_7[::-1]
print("Task 7:", reversed_arr_7, "\n")

# Task 8: Create a 2D array with 1 on the border and 0 inside
arr_8 = np.ones((5, 5))
arr_8[1:-1, 1:-1] = 0
print("Task 8:\n", arr_8, "\n")

# Task 9: Create a 8X8 matrix and fill it with a checkerboard pattern
arr_9 = np.zeros((8, 8), dtype=int)
arr_9[1::2, ::2] = 1
arr_9[::2, 1::2] = 1
print("Task 9:\n", arr_9, "\n")


Task 1: [10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33
 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49] 

Task 2:
 [[0 1 2]
 [3 4 5]
 [6 7 8]] 

Task 3:
 [[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
Task 4: Mean of array = 0.5256360842403772 

Task 5: Min = 0.01662442509477402 Max = 0.9833657933221076 

Task 6: [0. 0. 0. 0. 1. 0. 0. 0. 0. 0.] 

Task 7: [0 4 0 0 2 1] 

Task 8:
 [[1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 1.]
 [1. 0. 0. 0. 1.]
 [1. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1.]] 

Task 9:
 [[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]] 



Problem 3: Array Operations

In [13]:
import numpy as np

x = np.array([[1, 2], [3, 5]])
y = np.array([[5, 6], [7, 8]])
v = np.array([9, 10])
w = np.array([11, 12])

# 1. Add the two arrays
sum_xy = x + y
print("Addition of x and y:")
print(sum_xy, "\n")

# 2. Subtract the two arrays
subtract_xy = x - y
print("Subtraction of x and y:")
print(subtract_xy, "\n")

# 3. Multiply the arrays with any integers of your choice
multiply_x = x * 2
multiply_y = y * 2
print("Multiplication of x by 2:")
print(multiply_x)
print("Multiplication of y by 2:")
print(multiply_y, "\n")

# 4. Find the square of each element of the array
square_x = np.square(x)
square_y = np.square(y)
print("Square of each element in x:")
print(square_x)
print("Square of each element in y:")
print(square_y, "\n")

# 5. Dot product between v and w, x and v, x and y
dot_v_w = np.dot(v, w)
dot_x_v = np.dot(x, v)
dot_x_y = np.dot(x, y)

print("Dot product of v and w:", dot_v_w, "\n")
print("Dot product of x and v:", dot_x_v, "\n")
print("Dot product of x and y:", dot_x_y, "\n")

# 6. Concatenate x and y along rows and v and w along columns
concat_xy_rows = np.concatenate((x, y), axis=0)
print("Concatenate x and y along rows:")
print(concat_xy_rows, "\n")

concat_vw_cols = np.concatenate((v.reshape(-1, 1), w.reshape(-1, 1)), axis=1)
print("Concatenate v and w along columns:")
print(concat_vw_cols, "\n")

# 7. Concatenate x and v (Error case)
try:
    concat_x_v = np.concatenate((x, v), axis=0)
except ValueError as e:
    print("Error while concatenating x and v along rows:", e)


Addition of x and y:
[[ 6  8]
 [10 13]] 

Subtraction of x and y:
[[-4 -4]
 [-4 -3]] 

Multiplication of x by 2:
[[ 2  4]
 [ 6 10]]
Multiplication of y by 2:
[[10 12]
 [14 16]] 

Square of each element in x:
[[ 1  4]
 [ 9 25]]
Square of each element in y:
[[25 36]
 [49 64]] 

Dot product of v and w: 219 

Dot product of x and v: [29 77] 

Dot product of x and y: [[19 22]
 [50 58]] 

Concatenate x and y along rows:
[[1 2]
 [3 5]
 [5 6]
 [7 8]] 

Concatenate v and w along columns:
[[ 9 11]
 [10 12]] 

Error while concatenating x and v along rows: all the input arrays must have same number of dimensions, but the array at index 0 has 2 dimension(s) and the array at index 1 has 1 dimension(s)


Problem 4: Matrix Operations

In [14]:
import numpy as np

# Given matrices
A = np.array([[3, 4], [7, 8]])
B = np.array([[5, 3], [2, 1]])

# 1. Prove A * A_inv = I
A_inv = np.linalg.inv(A)
result = np.dot(A, A_inv)
print("A * A_inv =\n", result)

# 2. Prove AB != BA
AB = np.dot(A, B)
BA = np.dot(B, A)
print("AB =\n", AB)
print("BA =\n", BA)
print("Is AB equal to BA?", np.array_equal(AB, BA))

# 3. Prove (AB)^T = B^T * A^T
AB_T = np.transpose(np.dot(A, B))
B_T_A_T = np.dot(np.transpose(B), np.transpose(A))
print("(AB)^T =\n", AB_T)
print("B^T A^T =\n", B_T_A_T)
print("Is (AB)^T equal to B^T A^T?", np.array_equal(AB_T, B_T_A_T))


A * A_inv =
 [[1.00000000e+00 0.00000000e+00]
 [1.77635684e-15 1.00000000e+00]]
AB =
 [[23 13]
 [51 29]]
BA =
 [[36 44]
 [13 16]]
Is AB equal to BA? False
(AB)^T =
 [[23 51]
 [13 29]]
B^T A^T =
 [[23 51]
 [13 29]]
Is (AB)^T equal to B^T A^T? True


Solve the following system of Linear equation using Inverse Methods.

2x − 3y + z = −1

x − y + 2z = −3

3x + y − z = 9

In [ ]:
import numpy as np

A = np.array([[2, -3, 1],
              [1, -1, 2],
              [3, 1, -1]])

B = np.array([-1, -3, 9])

A_inv = np.linalg.inv(A)

X = np.dot(A_inv, B)

print("Solution (x, y, z) = ", X)


Solution (x, y, z) =  [ 2.  1. -2.]


Now: solve the above equation using np.linalg.inv function.{Explore more about ”linalg” func-
tion of Numpy}

In [ ]:
import numpy as np

A = np.array([[2, -3, 1],
              [1, -1, 2],
              [3, 1, -1]])

B = np.array([-1, -3, 9])

A_inv = np.linalg.inv(A)

X = np.dot(A_inv, B)

# Print the solution for x, y, and z
print("Solution (x, y, z) = ", X)

Solution (x, y, z) =  [ 2.  1. -2.]


# Compare plain python list with numpy list's operation

1. Element-wise Addition

In [17]:
import time

# Create two lists of size 1,000,000
list1 = [i for i in range(1000000)]
list2 = [i for i in range(1000000)]

# Measure time for element-wise addition using Python lists
start_time = time.time()

result_list = [list1[i] + list2[i] for i in range(1000000)]

end_time = time.time()
print(f"Time taken for element-wise addition using Python lists: {end_time - start_time:.6f} seconds")


Time taken for element-wise addition using Python lists: 0.255867 seconds


In [18]:
import numpy as np
import time

# Create two NumPy arrays of size 1,000,000
array1 = np.arange(1000000)
array2 = np.arange(1000000)

# Measure time for element-wise addition using NumPy arrays
start_time = time.time()

result_array = array1 + array2

end_time = time.time()
print(f"Time taken for element-wise addition using NumPy arrays: {end_time - start_time:.6f} seconds")


Time taken for element-wise addition using NumPy arrays: 0.012071 seconds


so, the total time taken using plain python is 0.255 second while numPy take 0.012 second.

2. Element-wise Multiplication

In [19]:
import time
import numpy as np

# Create two lists of size 1,000,000 with random numbers
list1 = [i for i in range(1, 1000001)]
list2 = [i for i in range(1, 1000001)]

# Measure the time taken for element-wise multiplication using Python lists
start_time = time.time()
result_list = [a * b for a, b in zip(list1, list2)]
list_time = time.time() - start_time

# Using Numpy arrays for the same operation
array1 = np.array(list1)
array2 = np.array(list2)

# Measure the time taken for element-wise multiplication using Numpy arrays
start_time = time.time()
result_array = array1 * array2
array_time = time.time() - start_time

print(f"Time taken with Python lists: {list_time} seconds")
print(f"Time taken with Numpy arrays: {array_time} seconds")


Time taken with Python lists: 0.29633092880249023 seconds
Time taken with Numpy arrays: 0.01678180694580078 seconds


so, numpy takes less (0.016) time then python plain list (0.296) operation 

3. Dot Product

In [1]:
import time
import numpy as np

# Using Python Lists
list1 = [i for i in range(1, 1000001)]
list2 = [i for i in range(1, 1000001)]

start_time = time.time()
dot_product_list = sum(list1[i] * list2[i] for i in range(1000000))
end_time = time.time()
print(f"Time taken using Python lists: {end_time - start_time} seconds")

# Using Numpy Arrays
array1 = np.arange(1, 1000001)
array2 = np.arange(1, 1000001)

start_time = time.time()
dot_product_numpy = np.dot(array1, array2)
end_time = time.time()
print(f"Time taken using NumPy arrays: {end_time - start_time} seconds")


Time taken using Python lists: 0.0948789119720459 seconds
Time taken using NumPy arrays: 0.0009639263153076172 seconds


so, the numpy is ahead in dot product too, numpy (0.0026) second and python list (0.357) second

4. Matrix Multiplication

In [2]:
import time
import numpy as np

# Matrix multiplication using Python lists

# Generate two 1000x1000 matrices with random values
matrix1 = [[1] * 1000 for _ in range(1000)]
matrix2 = [[1] * 1000 for _ in range(1000)]

# Function to multiply matrices using Python lists
def matrix_multiply_lists(A, B):
    size = len(A)
    result = [[0] * size for _ in range(size)]  # Resultant matrix
    for i in range(size):
        for j in range(size):
            result[i][j] = sum(A[i][k] * B[k][j] for k in range(size))
    return result

# Measure the time for matrix multiplication using Python lists
start_time = time.time()
result_lists = matrix_multiply_lists(matrix1, matrix2)
end_time = time.time()
time_lists = end_time - start_time
print(f"Time taken for matrix multiplication using Python lists: {time_lists:.5f} seconds")


# Matrix multiplication using NumPy

# Generate two 1000x1000 matrices with random values using NumPy
matrix1_np = np.ones((1000, 1000))
matrix2_np = np.ones((1000, 1000))

# Measure the time for matrix multiplication using NumPy
start_time = time.time()
result_np = np.dot(matrix1_np, matrix2_np)  # Alternatively, use matrix1_np @ matrix2_np
end_time = time.time()
time_np = end_time - start_time
print(f"Time taken for matrix multiplication using NumPy: {time_np:.5f} seconds")


KeyboardInterrupt: 

Huge difference between numpy and plain python list for multiplying matrix. 

Plain python matrix multiplication => 113.43 seconds

numpy python matrix multiplication => 0.072 second